clients/src/main/java/org/apache/kafka/common/security/authenticator/SaslClientAuthenticator.java

```diff
@@ -384,19 +475,33 @@ private boolean flushNetOutBuffer() throws IOException {
     }
 
     private AbstractResponse receiveKafkaResponse() throws IOException {
+        NetworkReceive receive = receivePacket();
+        if (receive == null)
+            return null;
+        byte[] responseBytes = new byte[receive.payload().remaining()];
+        receive.payload().get(responseBytes, 0, responseBytes.length);
+        /*
+         * Account for the fact that during re-authentication there may be responses
+         * arriving for requests that were sent in the past
+         */
         try {
-            byte[] responseBytes = receiveResponseOrToken();
-            if (responseBytes == null)
+            AbstractResponse response = NetworkClient.parseResponse(ByteBuffer.wrap(responseBytes), currentRequestHeader);
+            currentRequestHeader = null;
+            return response;
+        } catch (SchemaException | IllegalArgumentException e) {
+            if (reauthenticating) {
+                /*
+                 * It didn't match the current request header, so it must be unrelated to
+                 * re-authentication. Save it so it can be processed later.
+                 */
+                receive.payload().rewind();
+                pendingAuthenticatedReceives.add(receive);
```

Comment:
We have an exception that is going to result in the connection being closed. Seems unnecessary to re-authenticate before closing connection.

Reply:
Actually, my understanding of this is that we received something that had been sent prior to the re-authentication process beginning -- it doesn't match what we were expecting back, so we didn't send it, so it must have come from before -- so we have to save it so it can be processed later.  Does that make sense?  If so, then I think this code is correct.